core/kernel/tcg.c

```diff
@@ -0,0 +1,395 @@
+	if (tcg_get_pcr_info(&supported, &active, &pcr_count))
+		return TEE_ERROR_COMMUNICATION;
+
+	digest_list->count = 0;
+	/*
+	 * We have to take care that the sequence of algorithms that we record
+	 * in digest_list matches the sequence in eventlog.
+	 */
+	for (i = 0; i < alg_count; i++) {
+		hash_alg =
+		  get_unaligned_le16(&spec_event->digest_sizes[i].algorithm_id);
+
+		if (!(supported & tpm2_alg_to_tcg_mask(hash_alg))) {
+			EMSG("specID Event: Unsupported algorithm");
+			return TEE_ERROR_BAD_FORMAT;
+		}
+		digest_list->digests[digest_list->count++].hash_alg = hash_alg;
+
+		spec_active |= tpm2_alg_to_tcg_mask(hash_alg);
+	}
+
+	/*
+	 * TCG specification expects the event log to have hashes for all
+	 * active PCR's
+	 */
+	if (spec_active != active) {
+		/*
+		 * Previous stage bootloader should know all the active PCR's
+		 * and use them in the Eventlog.
+		 */
+		EMSG("specID Event: All active hash alg not present");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	/*
+	 * the size of the spec event and placement of vendor_info_size
+	 * depends on supported algoriths
+	 */
+	spec_event_size =
+		offsetof(struct tcg_efi_spec_id_event, digest_sizes) +
+		alg_count * sizeof(spec_event->digest_sizes[0]);
+
+	if (*pos + spec_event_size >= log_size)
+		return TEE_ERROR_BAD_FORMAT;
+
+	vendor_sz = *(uint8_t *)((uintptr_t)buffer + *pos + spec_event_size);
+
+	spec_event_size += sizeof(vendor_sz) + vendor_sz;
+	*pos += spec_event_size;
+
+	if (get_unaligned_le32(&event_header->event_size) != spec_event_size) {
+		EMSG("specID event: header event size mismatch");
+		/* Right way to handle this can be to call SetActive PCR's */
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * tcg_parse_event() -  Parse the event in the eventlog
+ *
+ * @buffer:		Pointer to the start of the eventlog
+ * @log_size:		Size of the eventlog
+ * @offset:		[in] Offset of the event in the eventlog buffer
+ *			[out] Return offset of the next event in the buffer
+ * @digest_list:	list of digests in the event
+ * @pcr			Index of the PCR in the event
+ *
+ * Return:		status code
+ */
+static TEE_Result tcg_parse_event(void *buffer, uint32_t log_size,
+				  uint32_t *offset,
+				  struct tpml_digest_values *digest_list,
+				  uint32_t *pcr)
+{
+	struct tcg_pcr_event2 *event = NULL;
+	uint32_t count = 0, size = 0, event_size = 0;
+	uint32_t i = 0;
+	size_t pos = 0;
+
+	event_size = tcg_event_final_size(digest_list);
+	if (*offset >= log_size || *offset + event_size > log_size) {
+		EMSG("Event exceeds log size");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	event = (struct tcg_pcr_event2 *)((uintptr_t)buffer + *offset);
+	*pcr = get_unaligned_le32(&event->pcr_index);
+
+	/* get the count */
+	count = get_unaligned_le32(&event->digests.count);
+	if (count != digest_list->count)
+		return TEE_ERROR_BAD_FORMAT;
+
+	pos = offsetof(struct tcg_pcr_event2, digests);
+	pos += offsetof(struct tpml_digest_values, digests);
+
+	for (i = 0; i < digest_list->count; i++) {
```

Comment:
What if `digest_list->count` isn't `TPM2_NUM_PCR_BANKS`, are we still accessing `struct tcg_pcr_event2` is it's defined in C or do we deviate?

Reply:
digest_lists count gets populated when parsing spec ID event,  Check that the count isn't TPM2_NUM_PCR_BANKS should happen there. I have added the check. Thanks for pointing this out.

There will be deviation from struct tcg_pcr_event2 , as actual size of digest (of type tcg_pcr_event2) would depend on the value of count which comes from spec ID event. That is the reason elements event_size and event are not accessed directly.

Reply:
So we have a struct definition that can't be used? That deserves at least a comment with a warning. We should also consider if there's a better way of documenting layout of binary data.